Added etfe function

AugustenFairbairn opened the pull request:
limited docstring that needs improvement, no tests avaliable

AugustenFairbairn committed:
Added etfe function

limited docstring that needs improvement, no tests avaliable

jsdodge reviewed (commented):
Hi @AugustenFairbairn , thanks for submitting this. I ran `hatch fmt` and it returned two errors, as follows:

```zsh
(thztools) (base) jsd@d207-023-184-217 thztools % hatch fmt
cmd [1] | ruff check --config '/Users/jsd/Library/Application Support/hatch/env/.internal/hatch-static-analysis/.config/gJvGbt9s/pyproject.toml' --fix .
error: Failed to parse src/thztools/thztools.py:3432:5: Expected ',', found name
src/thztools/thztools.py:3432:5: E999 SyntaxError: Expected ',', found name
src/thztools/thztools.py:3448:1: W293 Blank line contains whitespace
Found 2 errors.
No fixes available (1 hidden fix can be enabled with the `--unsafe-fixes` option).
```

The syntax error results from a missing comma at the end of line 3431. Please fix this and confirm that the following commands run wihout errors before resubmitting the PR.

```zsh
hatch fmt
hatch run test
hatch run docs:build
```

Here, the command `hatch fmt` will run the code through a linter and automatically reformat it according to [PEP8](https://peps.python.org/pep-0008/) standards. The command `hatch run test` will run the test suite, and `hatch run docs:build` will build the documentation locally.

AugustenFairbairn committed:
Fix etfe changes

fixed all the errors raised by hatch fmt, two errors still show but those are due to thz.fft not being pushed yet.

jsdodge reviewed (changes requested):
Hi @AugustenFairbairn , that's better, thanks! I still encounter some issues, so I'll bounce this back to you for additional changes.

The main issue is that the `fft` function is not defined in this code. I understand that @cmhchan was working on this, so it may be necessary to coordinate with her.
```zsh
(thztools) (base) jsd@d207-023-184-217 thztools % hatch fmt
cmd [1] | ruff check --config '/Users/jsd/Library/Application Support/hatch/env/.internal/hatch-static-analysis/.config/gJvGbt9s/pyproject.toml' --fix .
src/thztools/thztools.py:3449:11: F821 Undefined name `fft`
src/thztools/thztools.py:3450:11: F821 Undefined name `fft`
Found 2 errors.
```

Beyond that, I can see that there are some deviations from PEP8 formatting, which you should be able to fix automatically by running `hatch fmt` after including the `fft` code.

I recommend that you change `x_fft` and `y_fft` to `x_f` and `y_f`, respectively, for consistency with the rest of the code.

I don't think the `if... then` statement below is necessary. NumPy should issue a warning if it encounters a division by zero.
```python
    if np.any(x_fft == 0):
        msg = "fft of x contains zero values, leading to division by zero."
        raise ValueError(msg)
    else:
        xy_etfe = (y_fft/x_fft)
        return xy_etfe
```

AugustenFairbairn commented:
Hi @jsdodge, I'll make those consistency changes now, and use the fft function in Chloe's repo to make sure things will work correctly. Since fft isn't on the dev branch yet, It'll likely still return issues. Would you prefer that I hold off on pushing anything until the fft function has been merged first? Thanks!

AugustenFairbairn committed:
etfe formatting fixes

removed value error, changed variable conventions

jsdodge commented:
Yes, I think that it will probably be better to wait until the new `fft` function is implemented. We can also discuss any remaining issues when I meet with you in person.

AugustenFairbairn committed:
Updated etfe function to include the old fft function. Tests still need to be added.

AugustenFairbairn committed:
Updated etfe function, now accepts multidimensional arrays of different shapes. includes new argument that allows the user to choose the direction of the transform.

jsdodge commented:
Closed, superseded by #76
